File: 22-05-2024_Palindrome_Partitioning.cpp
class Solution {
public:
    
    int n ;

    bool is_palin(int a,int b,string &s){
        int i = a , j=b;
        while(i<=j){
            if(s[i]!=s[j]) return false;
            i++;
            j--;
        }

        return true;
    }

    void solve(int a , string &s , vector<vector<string>> &ans , vector<string> &temp){
        
        if(a==n){
            ans.push_back(temp);
            return;
        }

        for(int i=a;i<n;i++){
            temp.push_back(s.substr(a,i-a+1));
            if(is_palin(a,i,s))
            solve(i+1,s,ans,temp);
            temp.pop_back();
        }
    }

    vector<vector<string>> partition(string s) {
        n = s.size();
        vector<vector<string>> ans ;
        vector<string> temp;
        solve(0,s,ans,temp);

        return ans;
    }
};

File: 22-05-2024_Palindrome_Partitioning.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        self.n = len(s)
        self.ans = []
        self.solve(0, s, [])
        return self.ans

    def is_palin(self, a, b, s):
        i, j = a, b
        while i <= j:
            if s[i] != s[j]:
                return False
            i += 1
            j -= 1
        return True

    def solve(self, a, s, temp):
        if a == self.n:
            self.ans.append(temp[:])
            return

        for i in range(a, self.n):
            if self.is_palin(a, i, s):
                temp.append(s[a:i+1])
                self.solve(i + 1, s, temp)
                temp.pop()
